File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
// 자카드 유사도 교집합 / 합집합
// 교집합과 합집합 모두 0 이면 1
// 다중 집합에도 적용하자
// 다중 집합은 영어로만 확인한다.

import java.util.*;

class Solution {
    
    private static int NUMBER = 65536;
    
    public int solution(String str1, String str2) {
        int answer = 0;
        
        List<String> set1 = getMultiSet(str1);
        System.out.println(set1);
        List<String> set2 = getMultiSet(str2);
        System.out.println(set2);
        
        if (set1.isEmpty() && set2.isEmpty()) {
            answer = 1;
            return answer * NUMBER;
        }
        
        List<String> intersection = getIntersectionSet(set1, set2);
        System.out.println(intersection);
        List<String> union = getUnionSet(set1, set2);
        System.out.println(union);
        
        answer =  (int) Math.floor((intersection.size() / (float) union.size()) * NUMBER); 
        System.out.println(answer);
        
        return answer;
    }
    
    private List<String> getMultiSet(String str) {
        List<String> set = new ArrayList<>();
        for (int i = 0; i < str.length() - 1; i++) {
            String unit = str.substring(i, i + 2);
            if (unit.matches("^[a-zA-Z]*$")) {
                set.add(unit.toLowerCase());
            }
        }
        
        return set;
    }
    
    private List<String> getIntersectionSet(List<String> set1, List<String> set2) {
        int cnt = 0;
        List<String> intersectionSet = new ArrayList<>();
        for (String unit : set1) {
            if (set2.contains(unit) && !intersectionSet.contains(unit)) {
                int set1UnitCnt = getUnitCount(set1, unit);
                int set2UnitCnt = getUnitCount(set2, unit);
                cnt = Math.min(set1UnitCnt, set2UnitCnt);
                
                for (int i = 0; i < cnt; i++) {
                    intersectionSet.add(unit);
                }
            }
        }
        return intersectionSet;
    }
    
    private List<String> getUnionSet(List<String> set1, List<String> set2) {
        int cnt = 0;
        List<String> unionSet = new ArrayList<>();
        for (String unit : set1) {
            if (set2.contains(unit) && !unionSet.contains(unit)) {
                int set1UnitCnt = getUnitCount(set1, unit);
                int set2UnitCnt = getUnitCount(set2, unit);
                cnt = Math.max(set1UnitCnt, set2UnitCnt);
                
                for (int i = 0; i < cnt; i++) {
                    unionSet.add(unit);
                }
            } else if (!unionSet.contains(unit)) {
                int unitCnt = getUnitCount(set1, unit);
                for (int i = 0; i < unitCnt; i++) {
                    unionSet.add(unit);
                }
            }
        }
        
        for (String unit : set2) {
            if (!unionSet.contains(unit)) {
                int unitCnt = getUnitCount(set2, unit);
                for (int i = 0; i < unitCnt; i++) {
                    unionSet.add(unit);
                }
            }
        }
        return unionSet;
    }
    
    private int getUnitCount(List<String> set, String unit) {
        int cnt = 0;
        for (String setUnit : set) {
            if (setUnit.equals(unit)) {
                cnt++;
            }
        }
        return cnt;
    }
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
# 유사한 기사를 묶기 
# 자카드 유사도 : 집합간의 유사도를 검사하는 방법 
# 두 집합의 교집합의 크기를 합집합의 크기로 나눈 값
# 두 집합 모두 공집합일 경우 자카드 유사도는 1로 정의

# 중복을 허용하는 다중집합에 대해서 자카드 유사도를 구하기
# 문자열을 두글자씩 자른 배열 구하기 -> 영어만 포함된 경우 모두 소문자로 변경해서
# 교집합 구하기, 합집합 구하기
# 교집합의 길이를 합집합으로 나누고 65536 곱하기
from collections import Counter

def solution(str1, str2):
    answer = 0
    num = 65536
    
    str1 = makeArray(str1)
    str2 = makeArray(str2)
    
    if len(str1) == 0 and len(str2) == 0:
        return num
    
    dic1 = Counter(str1)
    dic2 = Counter(str2)
    
    sameUnit = 0
    allUnit = 0
    
    for unit in set(str1):
        if unit in dic1 and unit in dic2:
            sameUnit += min(dic1[unit], dic2[unit])
            allUnit += max(dic1[unit], dic2[unit])
        elif unit in dic1:
            allUnit += dic1[unit]
    for unit in set(str2):
        if unit in dic2 and unit not in dic1:
            allUnit += dic2[unit]
    
    jakade = sameUnit / allUnit
    answer = int(num * jakade)
    return answer

def makeArray(string):
    return [string[i : i + 2].lower() for i in range(0, len(string) - 1) if string[i : i + 2].isalpha()]
